From 6b28ad0675de7c4d4cd96d73c7049b6bcf191c35 Mon Sep 17 00:00:00 2001
From: Sujeet Baranwal <sbaranwal@marvell.com>
Date: Wed, 29 May 2019 16:58:33 -0700
Subject: [PATCH 0420/1921] mmc: octeontx2: cmd and data out values fixture

Cn95xx needs command and data out values to be fixed at 10
based on the HW recommendation.

Change-Id: Ib1a343a1fcce8645db4bc51ea5d534c2d899c181
Signed-off-by: Sujeet Baranwal <sbaranwal@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/10211
Reviewed-by: Sujeet Kumar Baranwal <Sujeet.Baranwal@cavium.com>
Reviewed-by: Chandrakala Chavva <cchavva@marvell.com>
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/10655
Tested-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/26915
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/mmc/host/cavium.c | 15 +++++++++------
 drivers/mmc/host/cavium.h | 10 ++++++++++
 2 files changed, 19 insertions(+), 6 deletions(-)

diff --git a/drivers/mmc/host/cavium.c b/drivers/mmc/host/cavium.c
index 0873a6932a78..f752056aff56 100644
--- a/drivers/mmc/host/cavium.c
+++ b/drivers/mmc/host/cavium.c
@@ -196,12 +196,15 @@ static int cvm_mmc_configure_delay(struct cvm_mmc_slot *slot)
 			break;
 		}
 
-		if (!cvm_is_mmc_timing_ddr(slot))
-			dout = cout;
-		else if (ddr_cmd_taps)
-			cout = dout = cout / 2;
-		else
-			dout = cout / 2;
+		if (!is_mmc_95xx(host)) {
+			if (!cvm_is_mmc_timing_ddr(slot))
+				dout = cout;
+			else if (ddr_cmd_taps)
+				cout = dout = cout / 2;
+			else
+				dout = cout / 2;
+		} else
+			dout = 10;
 
 		slot->taps =
 			FIELD_PREP(MIO_EMM_TIMING_CMD_IN, cin) |
diff --git a/drivers/mmc/host/cavium.h b/drivers/mmc/host/cavium.h
index c88e3de97baa..e757583d9848 100644
--- a/drivers/mmc/host/cavium.h
+++ b/drivers/mmc/host/cavium.h
@@ -28,6 +28,7 @@
 /* Subsystem Device ID */
 #define PCI_SUBSYS_DEVID_8XXX	0xA
 #define PCI_SUBSYS_DEVID_9XXX	0xB
+#define PCI_SUBSYS_DEVID_95XX	0xB3
 
 #define KHZ_400 (400000)
 #define MHZ_26  (26000000)
@@ -327,4 +328,13 @@ static inline bool is_mmc_otx2_A0(struct cvm_mmc_host *host)
 	return false;
 #endif
 }
+
+static inline bool is_mmc_95xx(struct cvm_mmc_host *host)
+{
+	struct pci_dev *pdev = host->pdev;
+	u32 chip_id = (pdev->subsystem_device >> 8) & 0xFF;
+
+	return (chip_id == PCI_SUBSYS_DEVID_95XX);
+}
+
 #endif
-- 
2.31.1

